### PR TITLE
Add debug options to push_firmware

### PR DESCRIPTION
**Pull request by Ircama**
Add debug options to `push_firmware`:

`-d|-dry-run`: Dry-run FTP commands without connecting to the device (for testing and development purposes).

`-s|-size <size>`: Force computing a specific file size for testing purpose. Imply dry-run.

Through these options, a developer can compare the parameters computed by `push_firmware` with the ones adopted by the AVM recovery tool in order to verify that for a specific model the AVM recovery tool parameters are correctly simulated.

After running the AVM Recovery Tool (https://download.avm.de/fritzbox/), the parameters used to reset the device and upload the firmware can be obtained via

```
notepad %localappdata%\VirtualStore\Windows\ftp.log
```

### CODE REVIEW

**Comment by fda77** [comment]
I dont see any usage for this. 
1) capture with tcpdump/wireshark the avm commands
2) write it to PF

so why should the commands in PF be debugged?
When avm adds a new 6th method to flash there is noting to debug in pf

**Comment by Ircama** [comment]
As a premise, I didn't want to annoy or create criticism with this PR (and with the others I created), but just to offer an help, respectfully.

The values from the recovery tool cannot be directly deducted, but must be calculated. The related process follows.

1. Download the recovery tool ([FRITZ.Box_7590_AX-07.57.exe](https://download.avm.de/fritzbox/fritzbox-7590-ax/other/recover/FRITZ.Box_7590_AX-07.57.exe) in my case)
2. Check the file size of the related image ([FRITZ.Box_7590_AX-07.57.image](https://download.avm.de/fritzbox/fritzbox-7590-ax/other/fritz.os/FRITZ.Box_7590_AX-07.57.image) in my case: 39915520 bytes = 0x2611000)
3. Read `notepad %localappdata%\VirtualStore\Windows\ftp.log`

4. From (2), FILESIZE should be 0x2611000 (39915520)

5. The output of (3) was:

  ```
  ...
  9:51:348: send: SETENV memsize 0x99fee00
  9:51:348: recv: 200 SETENV command successful
  9:51:348: send: SETENV kernel_args_tmp mtdram1=0x899fee00,0x8c000000
  ...
  9:51:352: send: STOR 0x899fee00 0x8c000000
  ...
  9:51:374: send image (size=39825920) for mtd1
  ...
  ```

6. Computations.

The target is to compute FREESIZE, MTDSTART and MAPLIMIT.

Analizing `STOR 0x899fee00 0x8c000000`, the actual file size could be 0x8c000000-0x899fee00 = 0x2601200 (39850496): there is a small difference: 0xFE00 = 65024. Let's accept it (also `send image (size=39825920) for mtd1` has a slightly different size) and suppose that the actual FILESIZE is 0x2601200 (39850496).

Analyzing *push_firmware*:

```
quote SETENV memsize FREESIZE
put fitimage MTDSTART MAPLIMIT
```

MAPLIMIT = 0x8c000000 (`STOR 0x899fee00 0x8c000000`)

MTDSTART = 0x899fee00 (`STOR 0x899fee00 0x8c000000`)

From `SETENV memsize 0x99fee00`, FREESIZE = 0x99fee00 (161476096).

Suppose ALIBYTES was 0, because the alignement was already normalized.

FREESIZE = FULLSIZE - FILESIZE - ALIBYTES => FULLSIZE = FREESIZE + FILESIZE + ALIBYTES = 0x99fee00 + 0x2601200 + 0 = 0xC000000 (201326592).

MTDSTART = MAPSTART + FULLSIZE - FILESIZE - ALIBYTES => MAPSTART = MTDSTART + FILESIZE + ALIBYTES - FULLSIZE = 0x899fee00 + 0x2601200 + 0 - 0xC000000 = 0x80000000 (2147483648).

In summary:
MAPSTART = 0x80000000
FULLSIZE = 0x0C000000

The `-d` and `-s` options help in verifying this.

Before consolidating the computed values, running the FTP manually and verifying that it works would also be a useful test.

**Comment by fda77** [comment]
You know your device 7590ax is not a FIT device? And you have no way to make it a FIT device? So forget FIT!
used memsize by recovery is sometime lower as real memsize. not sure why, maybe some areas are reserved and should not be used. Or maybe old defaults of avm. But it is sure 128mb (old default?) does not work always and it is sure the real ram size also does not work. So just watch recovery and get memsize+exe size and round it to ??000000... and the 0x80000000 offset is always for this flash method the same

> Before consolidating the computed values, running the FTP manually and verifying that it works would also be a useful test.

Its simple to know if the image in-the-memory boots - if it bootet or not

**Comment by Ircama** [comment]
> You know your device 7590ax is not a FIT device? And you have no way to make it a FIT device? So forget FIT!

Yes, that is another good point where I spent some time to check comments into this repository in order to understand whether 7590 AX supports the "fit-boot" mode. Fit is not an option for all devices, but the appropriate mode for specific ones which only support this mode because of the image format. Simply untar the image to understand whether it includes "fit-image".  Example (7530 AX is fit, 7590 AX is ram):

```
$ tar tvf FRITZ.Box_7530_AX-07.57.image
drwxr-xr-x 0/0               0 2023-12-20 13:19 ./var/
-rwxr-xr-x 0/0            1300 2023-12-20 13:19 ./var/install
drwxr-xr-x 0/0               0 2023-12-20 13:19 ./var/tmp/
-rw-r--r-- 0/0        40867336 2023-12-20 13:19 ./var/tmp/fit-image
-rw-r--r-- 0/0            2807 2023-12-20 13:19 ./var/info.txt
-rw-r--r-- 0/0              14 2023-12-20 13:19 ./var/install-features
-rw-r--r-- 0/0             356 2023-12-20 13:19 ./var/content
-rw-r--r-- 0/0              35 2023-12-20 13:19 ./var/version
-rw-r--r-- 0/0             128 2023-12-20 13:19 ./var/signature
```

`./var/tmp/fit-image` clarifies that such image for the 7530 AX is a "fit-boot" type.

Conversely, the 7590 AX image is a "ram-boot" type (and cannot be a "fit-boot"):

```
$ tar tvf FRITZ.Box_7590_AX-07.70-110138-LabBETA.image
drwxr-xr-x 0/0               0 2024-01-09 22:27 ./var/
-rw-r--r-- 0/0             356 2024-01-09 22:27 ./var/content
-rwxr-xr-x 0/0           55232 2024-01-09 22:27 ./var/chksum
-rw-r--r-- 0/0            2807 2024-01-09 22:27 ./var/info.txt
-rw-r--r-- 0/0              14 2024-01-09 22:27 ./var/install-features
drwxr-xr-x 0/0               0 2024-01-09 22:27 ./var/tmp/
-rw-r--r-- 0/0        33210376 2024-01-09 22:27 ./var/tmp/filesystem.image
-rw-r--r-- 0/0         7251208 2024-01-09 22:27 ./var/tmp/kernel.image
-rwxr-xr-x 0/0           20950 2024-01-09 22:27 ./var/install
-rw-r--r-- 0/0              42 2024-01-09 22:27 ./var/version
-rw-r--r-- 0/0             128 2024-01-09 22:27 ./var/signature
```

About the method you mentioned to heuristically attempt a number of possible values based on some assumptions until the firmware is uploaded to 7590 AX, appearing to be a fairly binding model, I always failed (I also tried reusing values valid for other RAM models). The only way I found that works is the one relating to the math I described above. That was the reason why I thought it useful to put some comments in this repo and propose some PR.

**Comment by fda77** [comment]
you know https://freetz-ng.github.io/freetz-ng/INSTALL.html ?

**Comment by Ircama** [comment]
> you know https://freetz-ng.github.io/freetz-ng/INSTALL.html ?

Yes, in fact that note was useful. But in the end I could not understand how to detect the actual mode adopted for 7590 AX, which I then tried to describe here below (e.g., untar).

**Comment by fda77** [comment]
Thats an internal thing you dont need to know, alien intelligence does it :)
With freetz you could check all devices with:  `grep FREETZ_AVM_HAS_FWLAYOUT_  config/.img/separate/*`

**Comment by Ircama** [comment]
Following the subsequent implementations on *push_firmware* that occurred meanwhile, I'd close this PR. We will eventually reopen it if it becomes useful.

**Comment by fda77** [comment]
I've put it currently to tools/push_firmware_experimental because some changes didnt work for other/older devices
